### PR TITLE
WASM package for berkeley qanet

### PR DESCRIPTION
**Pull request by sudo-shashank**
**Summary of changes**
Changes introduced in this pull request:
- Added new wasm package with api's for berkeley qanet

**Reference issue to close (if applicable)**

Closes 

**Other information and links**

### CODE REVIEW

File: apps/wasm/mina-berkeley/src/js_exports.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+#[wasm_bindgen]
+pub async fn poll_latest_blocks_once() -> Result<(), JsError> {
+    let backend = frontier::PROCESSOR_BERKELEY.nonconsensus_ops().await;
+    backend.poll_latest_once().await.map_err(err_to_js_error)
+}
+
+#[wasm_bindgen]
+pub async fn get_best_chain_state_hash() -> Option<String> {
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/js_exports.rs:35]: This api should return the protocol state

File: apps/wasm/mina-berkeley/src/js_exports.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+use crate::{logger::JsExportableLogger, *};
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+pub async fn run_processor() {
+    frontier::PROCESSOR_BERKELEY.run().await;
+}
+
+#[wasm_bindgen]
+pub async fn set_api_list(api_urls: String) -> Result<(), JsError> {
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/js_exports.rs:13]: Why not js array?

File: apps/wasm/mina-berkeley/src/js_exports.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    let mut backend = frontier::PROCESSOR_BERKELEY.nonconsensus_ops_mut().await;
+    let api_list: Vec<String> = api_urls.split('\n').map(|s| s.to_string()).collect();
+    backend.set_api_list(api_list.into_iter());
+    Ok(())
+}
+
+#[wasm_bindgen]
+pub async fn set_tracking_accounts(accounts: String) -> Result<(), JsError> {
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/js_exports.rs:21]: Ditto

File: apps/wasm/mina-berkeley/src/js_exports.rs (hunk 1 above)
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/js_exports.rs:35]: Need another api that returns the sparse merkle ledger

File: apps/wasm/mina-berkeley/src/js_exports.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    let mut backend = frontier::PROCESSOR_BERKELEY.nonconsensus_ops_mut().await;
+    let account_list: Vec<String> = accounts.split('\n').map(|s| s.to_string()).collect();
+    backend.set_tracking_accounts(account_list.into_iter());
+    Ok(())
+}
+
+#[wasm_bindgen]
+pub async fn poll_latest_blocks_once() -> Result<(), JsError> {
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/js_exports.rs:29]: There should be an api to run the process to kick off the event loop

**Review by hanabi1224 — CHANGES_REQUESTED** [review]

File: apps/wasm/mina-berkeley/src/js_exports.rs (hunk 4 above)
Comment (sudo-shashank) [reply apps/wasm/mina-berkeley/src/js_exports.rs:29]: this one?
https://github.com/ChainSafe/mina-rs/blob/7bbf4bd36e7dd4ec886c40666a60ed8ae8d42abb/apps/wasm/mina-berkeley/src/js_exports.rs#L8

Comment (hanabi1224) [reply apps/wasm/mina-berkeley/src/js_exports.rs:29]: Sry did not see that

File: apps/wasm/mina-berkeley/static/manifest.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
+{
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/static/manifest.json:1]: Files under static are for chrome extension v3, I don't think they're needed in this case.

File: apps/wasm/mina-berkeley/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsError;
+
+pub fn err_to_js_error<T: Display>(err: T) -> JsError {
+    JsError::new(err.to_string().as_str())
+}
+
+#[wasm_bindgen(typescript_custom_section)]
+const UNITS: &'static str = r#"
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/utils.rs:14]: What are UNITS for?

File: apps/wasm/mina-berkeley/src/js_exports.rs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    let resp_text = res.text().await.unwrap();
+    let resp: serde_json::Value = serde_json::from_str(&resp_text).unwrap();
+    Ok(resp.to_string())
+}
+
+#[wasm_bindgen(js_name = lockAccount)]
+pub async fn lock_account(pub_key: String) -> Result<String, JsError> {
+    let client = reqwest::Client::new();
+    let query: String = format!(
+        "mutation MyMutation {{
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/js_exports.rs:99]: What is this? Maybe move the impl to networking layer? How does this respect the api endpoint list if it uses hard-coded value

File: apps/wasm/mina-berkeley/src/js_exports.rs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+pub fn init_logger() -> Result<(), JsError> {
+    static JS_LOGGER: JsExportableLogger = JsExportableLogger::new(log::Level::Debug);
+    log::set_max_level(JS_LOGGER.max_level().to_level_filter());
+    log::set_logger(&JS_LOGGER).map_err(err_to_js_error)
+}
+
+#[wasm_bindgen(js_name = accountInfo)]
+pub async fn account_info(pub_key: String) -> Result<String, JsError> {
+    let client = reqwest::Client::new();
+    let query: String = format!(
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/js_exports.rs:78]: This info should come from sparse merkle ledger

File: apps/wasm/mina-berkeley/src/js_exports.rs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+#[wasm_bindgen(js_name = sendPayment)]
+pub async fn send_payment(payment: Payment) -> Result<String, JsError> {
+    let client = reqwest::Client::new();
+    let query: String = format!(
+        "mutation MyMutation {{  
+            sendPayment(input: {{fee: \"{}\", amount: \"{}\", to: \"{}\", from: \"{}\"}}) {{    payment {{      id    }}  }}
+        }}", payment.fee_u64()?, payment.amount_u64()?, payment.to(), payment.from()
+    );
+    let res = client
+        .post("http://localhost:3085/graphql")
```
Comment (hanabi1224) [inline apps/wasm/mina-berkeley/src/js_exports.rs:143]: Avoid hard-coding api endpoints

File: apps/wasm/mina-berkeley/src/js_exports.rs (hunk 7 above)
Comment (sudo-shashank) [reply apps/wasm/mina-berkeley/src/js_exports.rs:99]: locking and unlocking account is required before sending the payment

File: apps/wasm/mina-berkeley/src/js_exports.rs (hunk 8 above)
Comment (sudo-shashank) [reply apps/wasm/mina-berkeley/src/js_exports.rs:78]: okay I'll remove this

File: apps/wasm/mina-berkeley/src/js_exports.rs (hunk 9 above)
Comment (sudo-shashank) [reply apps/wasm/mina-berkeley/src/js_exports.rs]: fixed

**Review by hanabi1224 — APPROVED** [review]

**Review by creativcoder — APPROVED** [review]

**Pull request merged**